src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/FailedProxyCache.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// Holds a cache of failing proxies and manages when they should be retried.
+    /// </summary>
+    internal sealed class FailedProxyCache
+    {
+        /// <summary>
+        /// When returned by <see cref="GetProxyRenewTicks"/>, indicates a proxy is immediately usable.
+        /// </summary>
+        public const long Immediate = 0;
+
+        // If a proxy fails, time out 30 minutes. WinHTTP and Firefox both use this.
+        private const int FailureTimeoutInMilliseconds = 1000 * 60 * 30;
+
+        // Scan through the failures and flush any that have expired every 5 minutes.
+        private const int FlushFailuresTimerInMilliseconds = 1000 * 60 * 5;
+
+        // _failedProxies will only be flushed (rare but somewhat expensive) if we have more than this number of proxies in our dictionary. See Cleanup() for details.
+        private const int LargeProxyConfigBoundary = 8;
+
+        // Value is the Environment.TickCount64 to remove the proxy from the failure list.
+        private readonly ConcurrentDictionary<Uri, long> _failedProxies = new ConcurrentDictionary<Uri, long>();
+
+        // When Environment.TickCount64 >= _nextFlushTicks, cause a flush.
+        private long _nextFlushTicks = Environment.TickCount64 + FlushFailuresTimerInMilliseconds;
+
+        // This lock can be folded into _nextFlushTicks for space optimization, but
+        // this class should only have a single instance so would rather have clarity.
+        private SpinLock _flushLock = new SpinLock();
```

Comment:
@scalablecory, I only just noticed this.  Why did you choose to use a spin lock here rather than a regular lock?